File: backend/app/models/product.py
import uuid
from datetime import datetime
from decimal import Decimal
from typing import Optional, Any
from sqlalchemy import Boolean, DateTime, ForeignKey, Integer, Numeric, String, Text, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship
from app.database import Base


class BillingPlan(Base):
    __tablename__ = "billing_plans"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(50), nullable=False, unique=True)
    slug: Mapped[str] = mapped_column(String(30), nullable=False, unique=True)
    monthly_price_brl: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False, server_default="0")
    max_devices: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    max_users: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    ai_token_quota: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    sla_target_pct: Mapped[Optional[Decimal]] = mapped_column(Numeric(5, 2), nullable=True)
    features: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    subscriptions: Mapped[list["BillingSubscription"]] = relationship("BillingSubscription", back_populates="plan")


class BillingSubscription(Base):
    __tablename__ = "billing_subscriptions"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False, unique=True)
    plan_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("billing_plans.id"), nullable=False)
    stripe_customer_id: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    stripe_subscription_id: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    status: Mapped[str] = mapped_column(String(30), nullable=False, server_default="active")
    current_period_start: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    current_period_end: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    cancel_at_period_end: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    trial_end: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

    plan: Mapped["BillingPlan"] = relationship("BillingPlan", back_populates="subscriptions")
    invoices: Mapped[list["BillingInvoice"]] = relationship("BillingInvoice", back_populates="subscription")


class BillingInvoice(Base):
    __tablename__ = "billing_invoices"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    subscription_id: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), ForeignKey("billing_subscriptions.id", ondelete="SET NULL"), nullable=True)
    stripe_invoice_id: Mapped[Optional[str]] = mapped_column(String(200), nullable=True, unique=True)
    amount_brl: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    status: Mapped[str] = mapped_column(String(20), nullable=False, server_default="draft")
    period_start: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    period_end: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    paid_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    due_date: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    invoice_pdf_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())

    subscription: Mapped[Optional["BillingSubscription"]] = relationship("BillingSubscription", back_populates="invoices")


class OnboardingChecklist(Base):
    __tablename__ = "onboarding_checklists"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    tenant_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False)
    step_add_device: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    step_run_snapshot: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    step_ask_agent: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    step_configure_alert: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    completed: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    skipped: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    completed_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class HelpArticle(Base):
    __tablename__ = "help_articles"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    title: Mapped[str] = mapped_column(String(300), nullable=False)
    slug: Mapped[str] = mapped_column(String(200), nullable=False, unique=True)
    category: Mapped[str] = mapped_column(String(50), nullable=False, server_default="general")
    persona: Mapped[Optional[str]] = mapped_column(String(30), nullable=True)
    content_md: Mapped[str] = mapped_column(Text, nullable=False)
    is_published: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    view_count: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    sort_order: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    created_by: Mapped[Optional[uuid.UUID]] = mapped_column(UUID(as_uuid=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())


class UserPreference(Base):
    __tablename__ = "user_preferences"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), nullable=False, unique=True)
    language: Mapped[str] = mapped_column(String(10), nullable=False, server_default="pt-BR")
    timezone: Mapped[str] = mapped_column(String(50), nullable=False, server_default="America/Sao_Paulo")
    theme: Mapped[str] = mapped_column(String(20), nullable=False, server_default="dark")
    notifications_enabled: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")
    onboarding_step: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    onboarding_completed: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    extra: Mapped[Optional[Any]] = mapped_column(JSONB, nullable=True)
    updated_at: Mapped[datetime] = mapped_column(DateTime, server_default=func.now(), onupdate=func.now())

File: frontend/src/api/product.ts
import apiClient from "./client";

export interface BillingPlan {
  id: string;
  name: string;
  slug: string;
  monthly_price_brl: number;
  max_devices: number | null;
  max_users: number | null;
  ai_token_quota: number | null;
  sla_target_pct: number | null;
  features: Record<string, boolean> | null;
  is_active: boolean;
}

export interface BillingSubscription {
  id: string;
  tenant_id: string;
  status: string;
  cancel_at_period_end: boolean;
  current_period_start: string | null;
  current_period_end: string | null;
  trial_end: string | null;
  created_at: string;
  plan: BillingPlan;
}

export interface BillingInvoice {
  id: string;
  amount_brl: number;
  status: string;
  period_start: string | null;
  period_end: string | null;
  paid_at: string | null;
  due_date: string | null;
  invoice_pdf_url: string | null;
  created_at: string;
}

export interface OnboardingChecklist {
  id: string;
  step_add_device: boolean;
  step_run_snapshot: boolean;
  step_ask_agent: boolean;
  step_configure_alert: boolean;
  completed: boolean;
  skipped: boolean;
  completed_at: string | null;
}

export interface HelpArticle {
  id: string;
  title: string;
  slug: string;
  category: string;
  persona: string | null;
  content_md: string;
  is_published: boolean;
  view_count: number;
  sort_order: number;
  created_at: string;
}

export interface UserPreferences {
  id: string;
  language: string;
  timezone: string;
  theme: string;
  notifications_enabled: boolean;
  onboarding_step: number;
  onboarding_completed: boolean;
}

const BASE = "/product";

export const productApi = {
  // Plans
  listPlans: () =>
    apiClient.get<BillingPlan[]>(`${BASE}/billing/plans`).then(r => r.data),
  seedPlans: () =>
    apiClient.post<BillingPlan[]>(`${BASE}/billing/plans/seed`).then(r => r.data),

  // Subscription
  getSubscription: () =>
    apiClient.get<BillingSubscription | null>(`${BASE}/billing/subscription`).then(r => r.data),
  startSubscription: (plan_slug = "starter") =>
    apiClient.post<BillingSubscription>(`${BASE}/billing/subscription/start`, null, { params: { plan_slug } }).then(r => r.data),

  // Invoices
  listInvoices: () =>
    apiClient.get<BillingInvoice[]>(`${BASE}/billing/invoices`).then(r => r.data),

  // Onboarding
  getChecklist: () =>
    apiClient.get<OnboardingChecklist>(`${BASE}/onboarding/checklist`).then(r => r.data),
  completeStep: (step: string) =>
    apiClient.post<OnboardingChecklist>(`${BASE}/onboarding/checklist/complete-step`, { step }).then(r => r.data),
  skipOnboarding: () =>
    apiClient.post<OnboardingChecklist>(`${BASE}/onboarding/checklist/skip`).then(r => r.data),

  // Help
  listArticles: (category?: string, persona?: string) =>
    apiClient.get<HelpArticle[]>(`${BASE}/help/articles`, { params: { ...(category ? { category } : {}), ...(persona ? { persona } : {}) } }).then(r => r.data),
  getArticle: (slug: string) =>
    apiClient.get<HelpArticle>(`${BASE}/help/articles/${slug}`).then(r => r.data),
  seedArticles: () =>
    apiClient.post<HelpArticle[]>(`${BASE}/help/articles/seed`).then(r => r.data),

  // Preferences
  getPreferences: () =>
    apiClient.get<UserPreferences>(`${BASE}/preferences`).then(r => r.data),
  updatePreferences: (data: { language?: string; timezone?: string; theme?: string; notifications_enabled?: boolean }) =>
    apiClient.patch<UserPreferences>(`${BASE}/preferences`, data).then(r => r.data),

  // Stripe Checkout
  createCheckout: (plan_id: string, success_url?: string, cancel_url?: string) =>
    apiClient.post<{ checkout_url: string }>(`${BASE}/billing/checkout`, {
      plan_id,
      ...(success_url ? { success_url } : {}),
      ...(cancel_url ? { cancel_url } : {}),
    }).then(r => r.data),

  // Invoice PDF URL
  invoicePdfUrl: (invoice_id: string) => `/api${BASE}/billing/invoices/${invoice_id}/pdf`,
};
